src/parser/validate.ts

```diff
@@ -99,6 +71,96 @@ export function validate(parse: {
     }
   }
 
+  async function resolveFlags(flags: FlagRelationship[]): Promise<Record<string, unknown>> {
+    const promises = flags.map(async flag => {
+      if (typeof flag === 'string') {
+        return [flag, parse.output.flags[flag]]
+      }
+
+      const result = await flag.when(parse.output.flags)
+      return result ? [flag.name, parse.output.flags[flag.name]] : null
+    })
+    const resolved = await Promise.all(promises)
+    return Object.fromEntries(resolved.filter(r => r !== null) as [string, unknown][])
+  }
+
+  async function validateExclusive(name: string, flags: FlagRelationship[]) {
+    const resolved = await resolveFlags(flags)
+    const keys = Object.keys(resolved).reduce((acc, key) => {
+      if (resolved[key]) acc.push(key)
+      return acc
+    }, [] as string[])
+
+    for (const flag of keys) {
+      // do not enforce exclusivity for flags that were defaulted
+      if (
+        parse.output.metadata.flags[flag] &&
+        parse.output.metadata.flags[flag].setFromDefault
+      )
+        continue
+      if (
+        parse.output.metadata.flags[name] &&
+        parse.output.metadata.flags[name].setFromDefault
+      )
+        continue
+      if (parse.output.flags[flag]) {
+        throw new CLIError(
+          `--${flag}=${parse.output.flags[flag]} cannot also be provided when using --${name}`,
+        )
+      }
+    }
+  }
+
+  function validateExactlyOne(name: string, exactlyOne: FlagRelationship[]) {
+    for (const flag of exactlyOne || []) {
+      const flagName = typeof flag === 'string' ? flag : flag.name
+      if (flagName !== name && parse.output.flags[flagName]) {
+        throw new CLIError(
+          `--${flagName} cannot also be provided when using --${name}`,
+        )
+      }
+    }
+  }
+
+  async function validateDependsOn(name: string, flags: FlagRelationship[]) {
+    const resolved = await resolveFlags(flags)
+    const foundAll = Object.values(resolved).every(Boolean)
+    if (!foundAll) {
+      const formattedFlags = Object.keys(resolved).map(f => `--${f}`).join(', ')
+      throw new CLIError(
+        `All of the following must be provided when using --${name}: ${formattedFlags}`,
+      )
+    }
+  }
+
+  async function validateSome(name: string, flags: FlagRelationship[]) {
+    const resolved = await resolveFlags(flags)
+    const foundAtLeastOne = Object.values(resolved).some(Boolean)
+    if (!foundAtLeastOne) {
+      const formattedFlags = Object.keys(resolved).map(f => `--${f}`).join(', ')
+      throw new CLIError(`One of the following must be provided when using --${name}: ${formattedFlags}`)
+    }
+  }
+
+  async function validateRelationships(name: string, flag: CompletableFlag<any>) {
+    if (!flag.relationships) return
+    for (const relationship of flag.relationships) {
```

Comment:
worth parallelizing?

Reply:
ux idea: as written, you'll get a failure on the first validation to throw.  For the user, they'll fix one error and then hit the next.

You could run this with Promise.allSettled and maybe display all the errors in the output.